Update architecture overview (#1304)
vdaas/vald: docs/overview/architecture.md

Some components in Vald use Kubernetes API to control the behavior of distributed vector indexes.
Before reading this document, you need to have some understanding of the basic idea of cloud-native architecture and Kubernetes.

### Technologies used by Vald

Vald is based on the following techologies.

- [Kubernetes](https://kubernetes.io/)

> To easily scale and manage Vald, we deploy and run Vald on [Kubernetes](https://kubernetes.io/).

Reviewer: ```suggestion
    To easily scale and manage Vald, we deploy and run Vald on [Kubernetes](https://kubernetes.io/).
```

Author: I will commit it later

    Vald takes all of the advantages of using Kuberenetes.

>     For more details please read the next section.

Author: ```suggestion
    For more details please read the [next section](#concept).
```

- [Helm](https://helm.sh/)

> Helm help you to deploy and configure Vald. 

Reviewer 2: ```suggestion
    Helm help you to deploy and configure Vald. 
```

    Vald contains multiple components and configuration�.
    Helm helps us to manage those manifest and provides a better and easy way to deploy and configure Vald.

- [NGT](https://github.com/yahoojapan/NGT)

    NGT is one of the core component of Vald.

>     NGT is a super fast vector search engine used by Vald to guarantee the performance of Vald.

Reviewer 2: ```suggestion
    NGT is a super fast vector search engine used by Vald to guarantee the high performance of Vald.
```

  - Easy to manage

    Vald can be deployed easily on your Kubernetes cluster by using Helm charts. The custom resources and custom controllers are useful to manage your Vald cluster.

## Basic Architecture

Vald is base on microservice, which means Vald is composited by multiple components, you can deploy part of the components to your cluster depending on your needs.
In this section, we will introduce the minimal architecture of Vald.

<img src="../../assets/docs/overview/minimal_architecture.png" />

We will introduce each component and why it is needed in Vald.

### Vald Agent

> Vald Agent is the core engine of Vald, the approximate nearest neighbor search engine. Agent-NGT uses yahoojapan/NGT as a core library.

Author: ```suggestion
Vald Agent is the core component of Vald, the approximate nearest neighbor search engine, and stores the graph tree construction on memory for indexing the vectors.
Agent-NGT uses [yahoojapan/NGT](https://github.com/yahoojapan/ngt) as a core library.
```

  - Easy to manage

    Vald can be deployed easily on your Kubernetes cluster by using Helm charts. The custom resources and custom controllers are useful to manage your Vald cluster.

## Basic Architecture

Vald is based on microservice, which means Vald is composited by multiple components, you can deploy part of the components to your cluster depending on your needs.
In this section, we will introduce the basic architecture of Vald.

> <img src="../../assets/docs/overview/minimal_architecture.png" />

Author: should we convert all png to svg?

> <img src="../../assets/docs/overview/minimal_architecture.png" />

Reviewer: Yes, Please :pray:

> <img src="../../assets/docs/overview/minimal_architecture.png" />

Author: updated :)

We will introduce each component and why it is needed in Vald.

### Vald Agent

> Vald Agent is the core component of Vald, the approximate nearest neighbor search engine, and stores the graph tree construction on memory for indexing the vectors.
> Agent-NGT uses [yahoojapan/NGT](https://github.com/yahoojapan/ngt) as a core library.

Reviewer 2: ```suggestion
Vald Agent is the core component of Vald, the approximate nearest neighbor search engine, and stores the graph tree construction on memory for indexing the vectors.
Vald Agent uses [yahoojapan/NGT](https://github.com/yahoojapan/ngt) as a core library.
```

### Vald LB Gateway

> Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald-Agent based on the resource usage of the Vald-Agent and the corresponding cluster node.

Reviewer 2: Which is correct, `Vald Agent` or `Vald-Agent`?
I think we need to unify the names like any other component.(`Vald LB Gateway`)

```suggestion
Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald Agent based on the resource usage of the Vald Agent and the corresponding cluster node.
```

> Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald Agent based on the resource usage of the Vald Agent and the corresponding cluster node.

Author: ```suggestion
Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald Agent based on the resource usage of the Vald Agent and the corresponding cluster node.
In addition, it summarizes the search results from each Vald Agent and return the final search result to the client.
```

> Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald Agent based on the resource usage of the Vald Agent and the corresponding cluster node.

Reviewer: `return` -> `returns`

> Vald LB Gateway is a gateway to load balance the user request and forward user request to the Vald Agent based on the resource usage of the Vald Agent and the corresponding cluster node.

Author: edited

### Vald Discoverer

> Vald Discoverer provides Vald Agent discovery service to discover active Vald Agent pods in the Kubernetes cluster.

Reviewer 2: I think `Vald Discoverer` is not just for the vald Agent.

```suggestion
Vald Discoverer provides Vald Agent discovery service to discover active Vald Agent pods in the Kubernetes cluster.
```

Author: @hlts2 what is it also for?

Reviewer 2: I think Vald Discoverer provides the ability to interact with the Kubernetes API to get resource usage for active pods and their corresponding nodes. so Vald Discoverer is not just for the vald Agent.
if  I made a misstake, I am sorry.

Author: how about this?
```suggestion
Vald Discoverer provides Vald Agent discovery service to discover active Vald Agent pod and node resources in the Kubernetes cluster.
```

Reviewer 2: yes! looks good to me! thank you

> It also retrieves the corresponding Vald Agent resource usage for Vald LB gateway to determine the priority of which Vald Agent handles the user request.

Author: ```suggestion
It also retrieves the corresponding Vald Agent resource usage including pod and node resource usage for Vald LB gateway to determine the priority of which Vald Agent handles the user request.
```

### Vald Index Manager

Vald Index Manager controls the timing of the indexing of the Vald Agent.

> The index is used to increase the performance of the search action.

Reviewer: Please let me know what you'd like to talk.
I could not understand completely.

Author: I explain why indexes is required in Vald.
Do you think it is needed?

Reviewer 2: I think `Vald Index Manager` is not the main component 🤔 
so we can delete this section 🤔 
if I made a mistake, I am sorry.

Author: I think it is required to let user understand the full image of vald. what do you think?

Reviewer 2: I see. 🤔 ok. it looks good

Author: @vankichi thank you for your comment. I think it is better to say that the searching will not work when indexing.
```suggestion
Since the search operation will no work during the Vald Agent index operation is running, thanks to controlling the timing of indexing by Vald Index Manager, index operation of Vald Agent can be triggered and controlled by Vald Index Manager intelligently.
```
